### PR TITLE
Revisit atomics API

### PR DESCRIPTION
**Pull request by dalg24**
Related to #7449 

Get rid of the volatile-qualified overloads in favor of a single version that is properly constrained.
I plan to propose deprecations as follow-up work.

### CODE REVIEW

File: core/src/Kokkos_Atomics_Desul_Wrapper.hpp [hunk 1]
```diff
@@ -53,197 +53,118 @@ inline const char* atomic_query_version() { return "KOKKOS_DESUL_ATOMICS"; }
...
-
-template<class T> KOKKOS_INLINE_FUNCTION
-void atomic_store(T* const dest, desul::Impl::dont_deduce_this_parameter_t<const T> val) { return desul::atomic_store(dest, val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
-
-template<class T> KOKKOS_INLINE_FUNCTION
-void atomic_assign(T* const dest, desul::Impl::dont_deduce_this_parameter_t<const T> val) { atomic_store(dest,val); }
+namespace Impl {
+template <class T>
+using not_deduced_atomic_t =
+    std::add_const_t<std::remove_volatile_t<type_identity_t<T>>>;
```
Comment (dalg24) [inline core/src/Kokkos_Atomics_Desul_Wrapper.hpp:59]: The adding the const-qualifier is to match `desul::Impl::dont_deduce_this_parameter_t<const T>`.
I don't think it is particularly useful.

File: core/src/Kokkos_Atomics_Desul_Wrapper.hpp [hunk 2]
```diff
@@ -53,197 +53,118 @@ inline const char* atomic_query_version() { return "KOKKOS_DESUL_ATOMICS"; }
...
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_max(T* ptr, Impl::not_deduced_atomic_t<T> val) { desul::atomic_max(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_min(T* ptr, Impl::not_deduced_atomic_t<T> val) { desul::atomic_min(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_mul(T* ptr, Impl::not_deduced_atomic_t<T> val) { desul::atomic_mul(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_div(T* ptr, Impl::not_deduced_atomic_t<T> val) { desul::atomic_div(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+// mod
```
Comment (dalg24) [inline core/src/Kokkos_Atomics_Desul_Wrapper.hpp:120]: Here and below marking those that are missing.  Planning to add them in a follow up.

File: core/src/Kokkos_Atomics_Desul_Wrapper.hpp [hunk 3]
```diff
@@ -53,197 +53,118 @@ inline const char* atomic_query_version() { return "KOKKOS_DESUL_ATOMICS"; }
...
+KOKKOS_INLINE_FUNCTION void load_fence()   { desul::atomic_thread_fence(desul::MemoryOrderAcquire(), KOKKOS_DESUL_MEM_SCOPE); }
+KOKKOS_INLINE_FUNCTION void store_fence()  { desul::atomic_thread_fence(desul::MemoryOrderRelease(), KOKKOS_DESUL_MEM_SCOPE); }
 
-KOKKOS_INLINE_FUNCTION
-void store_fence() { return desul::atomic_thread_fence(desul::MemoryOrderRelease(), KOKKOS_DESUL_MEM_SCOPE); }
+// load/store
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T,    T> atomic_load (T const* ptr)                              { return desul::atomic_load (const_cast<std::remove_volatile_t<T>*>(ptr),      desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+template<class T> KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_store(T* ptr, Impl::not_deduced_atomic_t<T> val) { return desul::atomic_store(const_cast<std::remove_volatile_t<T>*>(ptr), val, desul::MemoryOrderRelaxed(), KOKKOS_DESUL_MEM_SCOPE); }
+template<class T> KOKKOS_DEPRECATED_WITH_COMMENT("Use atomic_store() instead!") KOKKOS_FUNCTION Impl::enable_if_atomic_t<T, void> atomic_assign(T* ptr, Impl::not_deduced_atomic_t<T> val) { atomic_store(ptr, val); }
```
Comment (dalg24) [inline core/src/Kokkos_Atomics_Desul_Wrapper.hpp:77]: Didn't mean to push that change (the marking as deprecated)
Will withdraw it.

**Review by crtrott — APPROVED** [review]
Approve to let the Sandia CI run

**Review by nliber — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

**Pull request merged**